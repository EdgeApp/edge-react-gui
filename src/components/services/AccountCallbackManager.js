// @flow

import { type EdgeAccount, type EdgeCurrencyWallet } from 'edge-core-js'
import { watchSecurityAlerts } from 'edge-login-ui-rn'
import * as React from 'react'

import { updateExchangeRates } from '../../actions/ExchangeRateActions.js'
import { checkPasswordRecovery } from '../../actions/RecoveryReminderActions.js'
import { newTransactionsRequest, refreshTransactionsRequest } from '../../actions/TransactionListActions.js'
import { updateWalletLoadingProgress, updateWalletsRequest } from '../../actions/WalletActions.js'
import { useAsyncEffect } from '../../hooks/useAsyncEffect.js'
import { useWalletsSubscriber } from '../../hooks/useWalletsSubscriber.js'
import { useEffect, useState } from '../../types/reactHooks.js'
import { useDispatch } from '../../types/reactRedux.js'
import { type NavigationProp, useNavigation } from '../../types/routerTypes.js'
import { isReceivedTransaction, snooze } from '../../util/utils.js'
import { WcSmartContractModal } from '../modals/WcSmartContractModal.js'
import { Airship } from './AirshipInstance.js'

type Props = {
  account: EdgeAccount
}

// Tracks items that need a refresh:
type DirtyList = {
  rates: boolean,
  walletList: boolean,
  wallets: { [walletId: string]: EdgeCurrencyWallet }
}

const notDirty: DirtyList = {
  rates: false,
  walletList: false,
  wallets: {}
}

export function AccountCallbackManager(props: Props) {
  const { account } = props
  const dispatch = useDispatch()
  const [dirty, setDirty] = useState<DirtyList>(notDirty)
  const navigation: NavigationProp<'edge'> = useNavigation()

  // Helper for marking wallets dirty:
  function addWallet(wallet: EdgeCurrencyWallet) {
    setDirty(dirty => ({
      ...dirty,
      rates: true, // We might reconsider this behavior
      wallets: { ...dirty.wallets, [wallet.id]: wallet }
    }))
  }

  // Subscribe to the account:
  useEffect(() => {
    const cleanups = [
      account.watch('currencyWallets', () =>
        setDirty(dirty => ({
          ...dirty,
          walletList: true
        }))
      ),

      account.watch('loggedIn', () => {
        if (account.loggedIn === false) {
          Airship.clear()
          console.log('onLoggedOut')
        }
      }),

      watchSecurityAlerts(account, hasAlerts => {
        navigation.push('securityAlerts')
      }),

      account.rateCache.on('update', () =>
        setDirty(dirty => ({
          ...dirty,
          rates: true
        }))
      )
    ]

    return () => cleanups.forEach(cleanup => cleanup())
  }, [account])

  // Subscribe to each wallet that comes online:
  useWalletsSubscriber(account, wallet => {
    const cleanups = [
      wallet.watch('syncRatio', ratio => {
        dispatch(updateWalletLoadingProgress(wallet.id, ratio))
      }),

      wallet.on('newTransactions', transactions => {
        console.log(`${walletPrefix(wallet)}: onNewTransactions: ${transactions.map(tx => tx.txid).join(' ')}`)

        dispatch(refreshTransactionsRequest(wallet.id, transactions))
        dispatch(newTransactionsRequest(wallet.id, transactions))
        addWallet(wallet)

        // Check if password recovery is set up:
        const finalTxIndex = transactions.length - 1
        if (isReceivedTransaction(transactions[finalTxIndex])) {
          dispatch(checkPasswordRecovery())
        }
      }),

      wallet.on('transactionsChanged', transactions => {
        console.log(`${walletPrefix(wallet)}: onTransactionsChanged: ${transactions.map(tx => tx.txid).join(' ')}`)

        dispatch(refreshTransactionsRequest(wallet.id, transactions))
        addWallet(wallet)
      }),

      wallet.on('wcNewContractCall', obj => {
        const { dApp, payload, uri, walletId } = obj
        if (walletId == null) return
        Airship.show(bridge => <WcSmartContractModal bridge={bridge} walletId={walletId} dApp={dApp} payload={payload} uri={uri} />)
      }),

      // These ones defer their work until later:
      wallet.watch('balances', () => addWallet(wallet)),
      wallet.watch('enabledTokenIds', () => addWallet(wallet)),
      wallet.watch('name', () => addWallet(wallet))
    ]

    return () => cleanups.forEach(cleanup => cleanup())
  })

  // Do the expensive work with rate limiting:
  useAsyncEffect(async () => {
    setDirty(notDirty)

    // Update wallets:
    const walletIds = Object.keys(dirty.wallets)
    if (dirty.walletList) {
      // Update all wallets (hammer mode):
      console.log('Updating wallet list')
      await dispatch(updateWalletsRequest())
      await snooze(1000)
    } else if (walletIds.length > 0) {
      // Update individual wallets:
      console.log(`Updating wallets: ${walletIds.map(id => walletPrefix(dirty.wallets[id])).join(' ')}`)
      dispatch({
        type: 'UI/WALLETS/UPSERT_WALLETS',
        data: {
          wallets: walletIds.map(id => dirty.wallets[id])
        }
      })
      await snooze(1000)
    }

    // Update exchange rates:
    if (dirty.rates) {
      console.log('Updating exchange rates')
      await dispatch(updateExchangeRates())
      await snooze(1000)
    }
  }, [dirty])

  return null
}

function walletPrefix(wallet: EdgeCurrencyWallet): string {
  return `${wallet.currencyInfo.currencyCode}-${wallet.id.slice(0, 2)}`
}
