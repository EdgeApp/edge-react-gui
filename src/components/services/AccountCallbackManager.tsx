import { EdgeAccount, EdgeCurrencyWallet } from 'edge-core-js'
import { watchSecurityAlerts } from 'edge-login-ui-rn'
import * as React from 'react'

import { showBackupModal } from '../../actions/BackupModalActions'
import { updateExchangeRates } from '../../actions/ExchangeRateActions'
import { checkFioObtData } from '../../actions/FioActions'
import { showReceiveDropdown } from '../../actions/ReceiveDropdown'
import { checkPasswordRecovery } from '../../actions/RecoveryReminderActions'
import { updateWalletLoadingProgress, updateWalletsRequest } from '../../actions/WalletActions'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useWalletsSubscriber } from '../../hooks/useWalletsSubscriber'
import { stakeMetadataCache } from '../../plugins/stake-plugins/metadataCache'
import { useDispatch } from '../../types/reactRedux'
import { NavigationBase } from '../../types/routerTypes'
import { snooze } from '../../util/utils'
import { Airship } from './AirshipInstance'

interface Props {
  account: EdgeAccount
  navigation: NavigationBase
}

// Tracks items that need a refresh:
interface DirtyList {
  rates: boolean
  walletList: boolean
}

const notDirty: DirtyList = {
  rates: false,
  walletList: false
}

export function AccountCallbackManager(props: Props) {
  const { account, navigation } = props
  const dispatch = useDispatch()
  const [dirty, setDirty] = React.useState<DirtyList>(notDirty)

  // Helper for marking wallets dirty:
  function setRatesDirty() {
    setDirty(dirty => ({
      ...dirty,
      rates: true
    }))
  }

  // Subscribe to the account:
  React.useEffect(() => {
    const cleanups = [
      account.watch('currencyWallets', () =>
        setDirty(dirty => ({
          ...dirty,
          walletList: true
        }))
      ),

      account.watch('loggedIn', () => {
        if (!account.loggedIn) {
          Airship.clear()
          console.log('onLoggedOut')
        }
      }),

      watchSecurityAlerts(account, hasAlerts => {
        if (hasAlerts) {
          navigation.navigate('securityAlerts', {})
        }
      }),

      account.rateCache.on('update', () =>
        setDirty(dirty => ({
          ...dirty,
          rates: true
        }))
      )
    ]

    return () => cleanups.forEach(cleanup => cleanup())
  }, [account, navigation])

  // Subscribe to each wallet that comes online:
  useWalletsSubscriber(account, wallet => {
    const cleanups = [
      wallet.watch('syncRatio', ratio => {
        dispatch(updateWalletLoadingProgress(wallet.id, ratio))
      }),

      wallet.on('newTransactions', transactions => {
        for (const tx of transactions) {
          const txid = tx.txid.toLowerCase()
          const cacheEntries = stakeMetadataCache[txid]
          // Assign cached stake metadata
          if (cacheEntries != null) {
            cacheEntries.forEach(cacheEntry => {
              const { currencyCode, metadata } = cacheEntry
              if (tx.currencyCode !== currencyCode) return
              wallet.saveTx({ ...tx, metadata }).catch(err => console.warn(err))
            })

            delete stakeMetadataCache[txid]
          }
        }

        console.log(`${walletPrefix(wallet)}: onNewTransactions: ${transactions.map(tx => tx.txid).join(' ')}`)

        // Check for incoming FIO requests:
        const receivedTxs = transactions.filter(tx => !tx.isSend)
        if (receivedTxs.length > 0) dispatch(checkFioObtData(wallet, receivedTxs)).catch(err => console.warn(err))

        // Show the dropdown for the first transaction:
        const [firstReceive] = receivedTxs
        if (firstReceive != null) {
          dispatch(showReceiveDropdown(navigation, firstReceive))

          // Notify the user to consider backing up their account
          if (account.username == null) {
            showBackupModal({ navigation })
          }
        }

        // Check if password recovery is set up:
        const finalTxIndex = transactions.length - 1
        if (!transactions[finalTxIndex].isSend) {
          dispatch(checkPasswordRecovery(navigation))
        }
      }),

      wallet.on('transactionsChanged', transactions => {
        console.log(`${walletPrefix(wallet)}: onTransactionsChanged: ${transactions.map(tx => tx.txid).join(' ')}`)
      }),

      // These ones defer their work until later:
      wallet.watch('balances', () => setRatesDirty()),
      wallet.watch('enabledTokenIds', () => setRatesDirty())
    ]

    return () => cleanups.forEach(cleanup => cleanup())
  })

  // Do the expensive work with rate limiting:
  useAsyncEffect(
    async () => {
      setDirty(notDirty)

      // Update wallets:
      if (dirty.walletList) {
        // Update all wallets (hammer mode):
        console.log('Updating wallet list')
        await dispatch(updateWalletsRequest())
        await snooze(1000)
      }

      // Update exchange rates:
      if (dirty.rates) {
        console.log('Updating exchange rates')
        await dispatch(updateExchangeRates())
        await snooze(1000)
      }
    },
    [dirty],
    'AccountCallbackManager'
  )

  return null
}

function walletPrefix(wallet: EdgeCurrencyWallet): string {
  return `${wallet.currencyInfo.currencyCode}-${wallet.id.slice(0, 2)}`
}
