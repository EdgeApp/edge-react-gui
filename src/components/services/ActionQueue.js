// @flow

import { type EdgeAccount } from 'edge-core-js'

import { updateActionProgramState } from '../../controllers/action-queue/redux/actions'
import { executeActionProgram } from '../../controllers/action-queue/runtime'
import { type ActionQueueMap } from '../../controllers/action-queue/types'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useRef } from '../../types/reactHooks'
import { useDispatch, useSelector } from '../../types/reactRedux'

export const ActionQueue = () => {
  const dispatch = useDispatch()
  const account: EdgeAccount = useSelector(state => state.core.account)
  const queue: ActionQueueMap = useSelector(state => state.actionQueue.queue)
  const executingRef = useRef<{ [programId: string]: boolean }>({})

  // TODO: Fix sync server stale ActionQueue state overwrite on startup
  //
  // Initialization
  //

  // useAsyncEffect(async () => {
  //   if (account?.dataStore != null) {
  //     const store = makeActionQueueStore(account)
  //     const queue = await store.getActionQueueMap()
  //     dispatch({
  //       type: 'ACTION_QUEUE/LOAD_QUEUE',
  //       data: queue
  //     })
  //   }
  // }, [account, dispatch])

  //
  // Runtime
  //

  useAsyncEffect(async () => {
    if (queue != null) {
      const executing = executingRef.current
      const promises = Object.keys(queue)
        .filter(
          programId =>
            // Ignore running programs
            !executing[programId]
        )
        .map(async programId => {
          // Set program to running
          executing[programId] = true

          const { program, state } = queue[programId]
          const { nextState } = await executeActionProgram(account, program, state).catch((error: Error) => ({
            nextState: {
              ...state,
              effect: { type: 'done', error }
            }
          }))

          // Update program state
          dispatch(updateActionProgramState(nextState))

          // Unset program to running
          executing[programId] = false
        })

      await Promise.all(promises)
    }
  }, [queue])

  return null
}
