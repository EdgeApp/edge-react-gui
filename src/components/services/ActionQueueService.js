// @flow

import { type EdgeAccount } from 'edge-core-js'

import { makeActionQueueStore } from '../../controllers/action-queue/ActionQueueStore'
import { mockActionProgram } from '../../controllers/action-queue/mock'
import { updateActionProgramState } from '../../controllers/action-queue/redux/actions'
import { executeActionProgram } from '../../controllers/action-queue/runtime'
import { type ActionQueueMap, type ExecutionContext } from '../../controllers/action-queue/types'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useRef } from '../../types/reactHooks'
import { useDispatch, useSelector } from '../../types/reactRedux'

export const ActionQueueService = () => {
  const dispatch = useDispatch()
  const account: EdgeAccount = useSelector(state => state.core.account)
  const clientId: string = useSelector(state => state.core.context.clientId)
  const queue: ActionQueueMap = useSelector(state => state.actionQueue.queue)
  const executingRef = useRef<{ [programId: string]: boolean }>({})

  const executionContext: ExecutionContext = { account, clientId }

  //
  // Initialization
  //

  useAsyncEffect(async () => {
    if (account?.dataStore != null) {
      const store = makeActionQueueStore(account, clientId)
      const queue = await store.getActionQueueMap()
      dispatch({
        type: 'ACTION_QUEUE/LOAD_QUEUE',
        data: queue
      })
    }
  }, [account, dispatch])

  //
  // Runtime
  //

  useAsyncEffect(async () => {
    if (queue != null) {
      const executing = executingRef.current
      const promises = Object.keys(queue)
        .filter(programId => !executing[programId] && queue[programId].state.clientId === clientId)
        .map(async programId => {
          // Set program to running
          executing[programId] = true

          const { program, state } = queue[programId]

          if (program.mockMode) {
            const { nextState } = await mockActionProgram(account, program, state)
            dispatch(updateActionProgramState(nextState))
            executing[programId] = false
            return
          }

          const { nextState } = await executeActionProgram(executionContext, program, state).catch((error: Error) => {
            console.warn(new Error('Action Program Exception: ' + error.message))
            console.error(error)
            return {
              nextState: {
                ...state,
                effect: { type: 'done', error }
              }
            }
          })

          // Update program state
          dispatch(updateActionProgramState(nextState))

          // Unset program to running
          executing[programId] = false
        })

      await Promise.all(promises)
    }
  }, [queue])

  return null
}
