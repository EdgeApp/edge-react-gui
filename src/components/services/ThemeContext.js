// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
