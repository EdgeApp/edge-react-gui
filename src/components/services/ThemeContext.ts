import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
