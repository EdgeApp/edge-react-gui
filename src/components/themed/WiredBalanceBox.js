// @flow

import * as React from 'react'
import { TouchableOpacity } from 'react-native'

import { toggleAccountBalanceVisibility } from '../../actions/WalletListActions.js'
import { getSymbolFromCurrency } from '../../constants/WalletAndCurrencyConstants.js'
import { formatNumber } from '../../locales/intl.js'
import s from '../../locales/strings.js'
import { connect } from '../../types/reactRedux.js'
import { type GuiExchangeRates } from '../../types/types.js'
import { getTotalFiatAmountFromExchangeRates } from '../../util/utils.js'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'
import { EdgeText } from './EdgeText.js'

type StateProps = {
  showBalance: boolean,
  fiatAmount: number,
  defaultIsoFiat: string,
  exchangeRates: GuiExchangeRates
}

type DispatchProps = {
  toggleAccountBalanceVisibility: () => void
}

type Props = StateProps & DispatchProps & ThemeProps

export class BalanceBox extends React.PureComponent<Props> {
  render() {
    const { defaultIsoFiat, fiatAmount, showBalance, exchangeRates, theme } = this.props
    const fiatSymbol = defaultIsoFiat ? getSymbolFromCurrency(defaultIsoFiat) : ''
    const fiatCurrencyCode = defaultIsoFiat.replace('iso:', '')
    const formattedFiat = formatNumber(fiatAmount, { toFixed: 2 })
    const styles = getStyles(theme)

    // if there is no exchangeRates object, empty object, or object with zero values
    // $FlowFixMe it appears that Object.values may break flow
    const summation = (total: number, rate: number) => {
      if (isNaN(rate)) rate = 0
      return total + rate
    }
    const noExchangeRates = !Object.keys(exchangeRates).length || !Object.values(exchangeRates).reduce(summation)

    return (
      <TouchableOpacity onPress={this.props.toggleAccountBalanceVisibility} style={styles.balanceBoxContainer}>
        {showBalance && !noExchangeRates ? (
          <>
            <EdgeText style={styles.balanceHeader}>{s.strings.fragment_wallets_balance_text}</EdgeText>
            <EdgeText style={styles.balanceBody}>
              {fiatSymbol.length !== 1 ? `${formattedFiat} ${fiatCurrencyCode}` : `${fiatSymbol} ${formattedFiat} ${fiatCurrencyCode}`}
            </EdgeText>
          </>
        ) : (
          <EdgeText style={styles.showBalance}>{noExchangeRates ? s.strings.exchange_rates_loading : s.strings.string_show_balance}</EdgeText>
        )}
      </TouchableOpacity>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  balanceBoxContainer: {
    height: theme.rem(3.25)
  },
  balanceHeader: {
    fontSize: theme.rem(1),
    color: theme.secondaryText
  },
  balanceBody: {
    fontSize: theme.rem(1.5),
    fontFamily: theme.fontFaceMedium
  },
  showBalance: {
    fontSize: theme.rem(1.5),
    fontFamily: theme.fontFaceMedium
  }
}))

export const WiredBalanceBox = connect<StateProps, DispatchProps, {}>(
  state => {
    const { defaultIsoFiat } = state.ui.settings
    return {
      showBalance: state.ui.settings.isAccountBalanceVisible,
      fiatAmount: getTotalFiatAmountFromExchangeRates(state, defaultIsoFiat),
      defaultIsoFiat,
      exchangeRates: state.exchangeRates
    }
  },
  dispatch => ({
    toggleAccountBalanceVisibility() {
      dispatch(toggleAccountBalanceVisibility())
    }
  })
)(withTheme(BalanceBox))
