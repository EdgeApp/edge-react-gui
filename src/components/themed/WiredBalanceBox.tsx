import * as React from 'react'
import { TouchableOpacity } from 'react-native'

import { toggleAccountBalanceVisibility } from '../../actions/LocalSettingsActions'
import { getSymbolFromCurrency } from '../../constants/WalletAndCurrencyConstants'
import { formatNumber } from '../../locales/intl'
import { lstrings } from '../../locales/strings'
import { connect } from '../../types/reactRedux'
import { GuiExchangeRates } from '../../types/types'
import { triggerHaptic } from '../../util/haptic'
import { getTotalFiatAmountFromExchangeRates } from '../../util/utils'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../services/ThemeContext'
import { EdgeText } from './EdgeText'
import { SceneHeader } from './SceneHeader'

interface StateProps {
  showBalance: boolean
  fiatAmount: number
  defaultIsoFiat: string
  exchangeRates: GuiExchangeRates
}

interface DispatchProps {
  toggleAccountBalanceVisibility: () => void
}

type Props = StateProps & DispatchProps & ThemeProps

export class BalanceBox extends React.PureComponent<Props> {
  handleToggleAccountBalanceVisibility = () => {
    triggerHaptic('impactLight')
    this.props.toggleAccountBalanceVisibility()
  }

  render() {
    const { defaultIsoFiat, fiatAmount, showBalance, exchangeRates, theme } = this.props
    const fiatSymbol = defaultIsoFiat ? getSymbolFromCurrency(defaultIsoFiat) : ''
    const fiatCurrencyCode = defaultIsoFiat.replace('iso:', '')
    const formattedFiat = formatNumber(fiatAmount, { toFixed: 2 })
    const styles = getStyles(theme)

    // if there is no exchangeRates object, empty object, or object with zero values

    const summation = (total: number, rate: number) => {
      if (isNaN(rate)) rate = 0
      return total + rate
    }
    // @ts-expect-error
    const noExchangeRates = !Object.keys(exchangeRates).length || !Object.values(exchangeRates).reduce(summation)

    return (
      <SceneHeader underline>
        <TouchableOpacity onPress={this.handleToggleAccountBalanceVisibility} style={styles.balanceBoxContainer}>
          {showBalance && !noExchangeRates ? (
            <>
              <EdgeText style={styles.balanceHeader}>{lstrings.fragment_wallets_balance_text}</EdgeText>
              <EdgeText style={styles.balanceBody}>
                {fiatSymbol.length !== 1 ? `${formattedFiat} ${fiatCurrencyCode}` : `${fiatSymbol} ${formattedFiat} ${fiatCurrencyCode}`}
              </EdgeText>
            </>
          ) : (
            <EdgeText style={styles.showBalance}>{noExchangeRates ? lstrings.exchange_rates_loading : lstrings.string_show_balance}</EdgeText>
          )}
        </TouchableOpacity>
      </SceneHeader>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  balanceBoxContainer: {
    height: theme.rem(3.25),
    marginTop: theme.rem(0.5)
  },
  balanceHeader: {
    fontSize: theme.rem(1),
    color: theme.secondaryText
  },
  balanceBody: {
    fontSize: theme.rem(1.5),
    fontFamily: theme.fontFaceMedium
  },
  showBalance: {
    fontSize: theme.rem(1.5),
    fontFamily: theme.fontFaceMedium
  }
}))

export const WiredBalanceBox = connect<StateProps, DispatchProps, {}>(
  state => {
    const { defaultIsoFiat } = state.ui.settings
    return {
      showBalance: state.ui.settings.isAccountBalanceVisible,
      fiatAmount: getTotalFiatAmountFromExchangeRates(state, defaultIsoFiat),
      defaultIsoFiat,
      exchangeRates: state.exchangeRates
    }
  },
  dispatch => ({
    toggleAccountBalanceVisibility() {
      dispatch(toggleAccountBalanceVisibility())
    }
  })
)(withTheme(BalanceBox))
