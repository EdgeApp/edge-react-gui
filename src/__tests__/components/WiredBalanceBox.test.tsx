/* globals describe it expect */

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { getTheme } from '../../components/services/ThemeContext'
import { BalanceBox } from '../../components/themed/WiredBalanceBox'

describe('BalanceBox', () => {
  it('should render with loading props', () => {
    // @ts-expect-error
    const renderer = new ShallowRenderer()

    const props = {
      showBalance: true,
      fiatAmount: 11,
      defaultIsoFiat: 'string',
      exchangeRates: 'GuiExchangeRates',
      toggleAccountBalanceVisibility: () => undefined,
      theme: getTheme()
    }
    // @ts-expect-error
    const actual = renderer.render(<BalanceBox {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
