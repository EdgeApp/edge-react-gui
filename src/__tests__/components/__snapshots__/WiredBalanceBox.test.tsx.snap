// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BalanceBox should render with loading props 1`] = `
<Memo(SceneHeaderComponent)
  underline={true}
>
  <TouchableOpacity
    onPress={[Function]}
    style={
      Object {
        "height": 73,
      }
    }
  >
    <React.Fragment>
      <WithTheme(EdgeTextComponent)
        style={
          Object {
            "color": "#A4C7DF",
            "fontSize": 22,
          }
        }
      >
        Total Balance
      </WithTheme(EdgeTextComponent)>
      <WithTheme(EdgeTextComponent)
        style={
          Object {
            "fontFamily": "Quicksand-Medium",
            "fontSize": 34,
          }
        }
      >
        11.00 string
      </WithTheme(EdgeTextComponent)>
    </React.Fragment>
  </TouchableOpacity>
</Memo(SceneHeaderComponent)>
`;
